src/librustc_metadata/encoder.rs

```diff
@@ -116,12 +117,32 @@ impl<'a, 'tcx> serialize::Encoder for ::encoder::EncodeContext<'a, 'tcx> {
 
 impl<'a, 'tcx> SpecializedEncoder<Ty<'tcx>> for EncodeContext<'a, 'tcx> {
     fn specialized_encode(&mut self, ty: &Ty<'tcx>) -> Result<(), Self::Error> {
-        let cx = self.ty_str_ctxt();
+        let existing_shorthand = self.type_shorthands.borrow().get(ty).cloned();
+        if let Some(shorthand) = existing_shorthand {
+            return self.emit_usize(shorthand);
+        }
+
+        let start = self.mark_stable_position();
+        ty.sty.encode(self)?;
+        let len = self.mark_stable_position() - start;
+
+        // The shorthand encoding uses the same usize as the
+        // discriminant, with an offset so they can't conflict.
+        let discriminant = unsafe { intrinsics::discriminant_value(&ty.sty) };
+        assert!(discriminant < TYPE_SHORTHAND_OFFSET as u64);
```

Comment:
This assert is good, I guess, it'd be nice if we could assert the shorthand-encoding trick *just a bit* more directly -- but it seems ok I guess.